docs/lifecycle-management.md

```diff
@@ -0,0 +1,225 @@
+So far KubeVirt provides a pod-like interface when managing VMs. This seems to
+be a good building block.
+On the other hand this might not be sufficient for usual virtualization
+management UIs, which do expect to manage the complete life-cycle of a VM,
+including shut down VMs (which is currently not intended by the core runtime).
+
+
+This document outlines how minimal assumptions can be met and also visits how
+other requirements impact the design.
+
+
+## Requirements
+* Running and non-running VMs are handled through the same API
+* Creating a VM must be separate from launching a VM
+* Support for “flavor” based VMs (templating)
+* Support for expanding sparse VM definitions
+* Support for detailed VM definitions
+* Support hot-(un-)plugging
+* Representation of dynamic instance specific data
+* Support for defaulting (to support OS compatibility)
+Considerations
+* Sub-resources for actions
+* Presets to add certain configurations
+
+
+## Definitions
+* **Defaulting** - Setting certain defaults during expansion (i.e. virtio for a
+  bus type, or e1000 for as a nic model)
+* **Expansion** - VM definitions can be provided in a sparse way i.e. just a
+  disk. Once the VM is getting started, the runtime will infer additional
+  requirements for this device and add them to the runtime VM definition.
+  This process of inferring and adding additional devices is called expansion.
+* **Flavor** - Comparable to a template or a specific instance type
+* **Preset** - Presetting certain values in a specified path of the document
+  (i.e. setting .spec.domain.devices.nvram). It is not about setting certain
+  default values for undefined fields (i.e. set virtio for all busses)
+
+
+## Option ICID - “Imperative-Config-Instance-Dance”
+### Kinds
+
+**Template/Config**
+```yaml
+Kind: Config
+Metadata:
+  # If a config for a started instance should be created
+  instanceConfig: False
+  # If instanceConfig: Automatically save changes to the instance to the
+  # associated Config
+  autoSave: False
+Spec:
+  Template:
+    Metadata:
+    Spec:
+      Domain:
+        Devices:
+          …
+```
+
+**Instance**
+```yaml
+Kind: VM
+Metadata:
+  ownerReferences: [{$config}]
+Spec:
+  Domain:
+    Devices:
+      …
+State:
+  Domain:
+    …
+```
+
+### Flow
+
+```
+$c1        $c2        $c3        $i1        $i2        $i3       $i4
+POST
+/start                           POST
+           POST
+/stop                            DELETE
+           /start                           POST
+           /snapshot  POST
+           /save PATCH
+           /stop                            DELETE
+           /start                                      POST
+           PATCH                    ←                  PATCH
+           PATCH
+PATCH
+/start
+                                                                 POST
+/stop                                                            DELETE
+```
+
+**Note:** The `/snapshot` call is just an exmaple how this mechanism could
```

Comment:
I think we could completely remove the snapshot endpoint, it can be modelled imperative and there is a good chance that it is 100% runtime dependent, not config dependent.